ntp-os-clock/src/unix.rs

```diff
@@ -0,0 +1,439 @@
+};
+
+#[cfg(any(target_os = "freebsd", target_os = "macos"))]
+pub(crate) const EMPTY_TIMEX: libc::timex = libc::timex {
+    modes: 0,
+    offset: 0,
+    freq: 0,
+    maxerror: 0,
+    esterror: 0,
+    status: 0,
+    constant: 0,
+    precision: 0,
+    tolerance: 0,
+    ppsfreq: 0,
+    jitter: 0,
+    shift: 0,
+    stabil: 0,
+    jitcnt: 0,
+    calcnt: 0,
+    errcnt: 0,
+    stbcnt: 0,
+};
+
+/// NTP Clock that uses the unix NTP KAPI clock functions to get/modify the
+/// current time.
+// Implementation note: this is intentionally a bare struct, the NTP Clock defined
+// in the NTP KAPI is unique and no state is needed to interact with it.
+#[derive(Debug, Default, Clone)]
+pub struct UnixNtpClock {
+    clock: libc::clockid_t,
+}
+
+impl UnixNtpClock {
+    pub fn realtime() -> Self {
+        Self::custom(libc::CLOCK_REALTIME)
+    }
+
+    pub fn custom(id: libc::clockid_t) -> Self {
+        Self { clock: id }
+    }
+
+    #[cfg_attr(target_os = "linux", allow(unused))]
+    fn clock_gettime(&self) -> Result<libc::timespec, Error> {
+        let mut timespec = libc::timespec {
+            tv_sec: 0,
+            tv_nsec: 0,
+        };
+
+        // # Safety
+        //
+        // using an invalid clock id is safe. `clock_adjtime` will return an EINVAL error
+        // https://linux.die.net/man/3/clock_gettime
+        //
+        // The timespec pointer is valid.
+        cerr(unsafe { libc::clock_gettime(self.clock, &mut timespec) })?;
+
+        Ok(timespec)
+    }
+
+    #[cfg_attr(target_os = "linux", allow(unused))]
+    fn clock_settime(&self, mut timespec: libc::timespec) -> Result<(), Error> {
+        while timespec.tv_nsec > 1_000_000_000 {
+            timespec.tv_sec += 1;
+            timespec.tv_nsec -= 1_000_000_000;
+        }
+
+        // # Safety
+        //
+        // using an invalid clock id is safe. `clock_adjtime` will return an EINVAL error
+        // https://linux.die.net/man/3/clock_settime
+        //
+        // The timespec pointer is valid.
+        unsafe { cerr(libc::clock_settime(self.clock, &timespec))? };
+
+        Ok(())
+    }
+
+    fn clock_adjtime(&self, timex: &mut libc::timex) -> Result<(), Error> {
+        // We don't care about the time status, so the non-error
+        // information in the return value of clock_adjtime can be ignored.
+        //
+        // # Safety
+        //
+        // The clock_adjtime call is safe because the reference always
+        // points to a valid libc::timex.
+        //
+        // using an invalid clock id is safe. `clock_adjtime` will return an EINVAL error
+        // https://man.archlinux.org/man/clock_adjtime.2.en#EINVAL~4
+        #[cfg(target_os = "linux")]
+        use libc::clock_adjtime as adjtime;
+
+        #[cfg(any(target_os = "freebsd", target_os = "macos"))]
+        let adjtime = {
+            extern "C" {
+                fn clock_adjtime(clk_id: libc::clockid_t, buf: *mut libc::timex) -> libc::c_int;
+            }
```

Comment:
based on https://www.gnu.org/software/gnulib/manual/html_node/clock_005fadjtime.html this function should exist on modern macos/freebsd. I'll see if we can get that upstreamed into libc

Reply:
but this is not so relevant here (or for hardware timestamping)

Reply:
This seems not to exist looking at the freebsd man pages, so we should be careful with this once we start actually testing that.